Fix #19123 - reCaptcha site verify URL

xboxfly15 opened the pull request:
From my mentioned issue a few months ago, I've decided to submit a PR myself.
I came across a bug while using reCaptcha site verify URL, this PR moves where the site verify URL is checked, to where reCaptcha now expects it to be.

I followed this reCaptcha documentation: https://github.com/google/recaptcha/blob/master/ARCHITECTURE.md https://github.com/google/recaptcha/tree/master?tab=readme-ov-file#usage

Fixes #19123

xboxfly15 committed:
Fix reCaptcha site verify URL
Signed-off-by: xboxfly15 <19935166+xboxfly15@users.noreply.github.com>

williamdes commented on src/Plugins/Auth/AuthenticationCookie.php:
```diff
@@ -248,22 +248,22 @@ public function readCredentials(): bool
 
                 // verify captcha status.
-                $resp = $reCaptcha->verify(
+                $resp = $reCaptcha->setExpectedHostname($captchaSiteVerifyURL)->verify(
```
I guess you need to add a null check from what our tools are reporting

williamdes commented:
> From my mentioned issue a few months ago, I've decided to submit a PR myself.

Thank you, this is a good idea. And sorry for not sorting out and labeling the issue you posted earlier.

xboxfly15 committed:
Revert my mistake of removing custom siteVerifyUrl, add captchaExpectedHostname and improve documentation.
This is the proper, correct fix for #19123
Signed-off-by: xboxfly15 <19935166+xboxfly15@users.noreply.github.com>

xboxfly15 commented:
Thank you for pointing out needing the null check, it made me take a second look at everything and I have since looked at more reCaptcha documentation and when you added `CaptchaSiteVerifyURL` 4 years ago in commit 3278415

I do apologise, I was mistaken between the difference of `captchaSiteVerifyURL` and expected hostname. 
I was treating `captchaSiteVerifyURL` as if it was the expected hostname, when it is in fact not, I most likely made this mistake due to poor phpMyAdmin documentation.

I have corrected my mistake, improved the `captchaSiteVerifyURL` documentation, added a new configuration variable `CaptchaExpectedHostname` that fixes my initial issue (with documentation and null check), this new variable improves reCaptcha security as stated https://developers.google.com/recaptcha/docs/domain_validation#security_warning

The only thing I am unsure about is in the documentation, for `captchaExpectedHostname` I have done `versionadded:: 5.2.0`, I am not sure what version should be there, so just assumed.

xboxfly15 committed:
Merge branch 'phpmyadmin:master' into fix-recaptcha

xboxfly15 commented:
Apologises for merging, I got a bit carried away and forgot haha, happy to recreate PR.
I have been doing testing this all weekend. I have tested this change (using v2 invisible and v2 checkbox) and it works! "Verify the origin of reCAPTCHA solutions" was disabled in the site console for all 3 tests.
- Having no `CaptchaExpectedHostname` set, lets you login still - as it did before this change (this is the security flaw that [reCAPTCHA warns about on their developer site](https://developers.google.com/recaptcha/docs/domain_validation))
- Having `CaptchaExpectedHostname` set to the wrong hostname or trying to login via different URL than the captcha is expecting, fails and shows the error "Entered captcha is wrong, try again!"
- Having `CaptchaExpectedHostname` set to the correct hostname and/or using the correct URL matching the value, works and lets you login.

The `We detected that your site isn't verifying reCAPTCHA solutions.` warning still shows but can be safely ignored when `CaptchaExpectedHostname` is set.


_@williamdes I just want to make you aware, using `auth_type=config` for a server, captchas are not verified server-side at all. This seems quite concerning to me and has caused me a lot of confusion over the weekend, while I was testing.
The captcha still shows up and I get the warning `We detected that your site isn't verifying reCAPTCHA solutions. This is required for the proper use of reCAPTCHA on your site. Please see our developer site for more information.` in the recaptcha console_

williamdes commented on src/Config/Descriptions.php:
```diff
@@ -951,6 +952,7 @@ private static function getNames(): array
             'CaptchaLoginPublicKey_name' => __('Public key for reCAPTCHA'),
             'CaptchaLoginPrivateKey_name' => __('Private key for reCAPTCHA'),
             'CaptchaSiteVerifyURL_name' => __('URL for reCAPTCHA siteverify'),
+            'CaptchaExpectedHostname_name' => __('Domain for reCAPTCHA to ensure captcha came from phpMyAdmin'),
```
```suggestion
            'CaptchaExpectedHostname_name' => __('Domain for reCAPTCHA to ensure the captcha request came from phpMyAdmin'),
```

williamdes commented on docs/config.rst:
```diff
@@ -2158,12 +2158,28 @@ Cookie authentication options
+
+    reCaptcha will be then used in :ref:`cookie`.
+
+    .. versionadded:: 5.2.0
```
```suggestion
    .. versionadded:: 5.2.2
```

xboxfly15 committed:
Update src/Config/Descriptions.php

Co-authored-by: William Desportes <williamdes@wdes.fr>

xboxfly15 committed:
Update docs/config.rst

Co-authored-by: William Desportes <williamdes@wdes.fr>

xboxfly15 commented:
Cheers! 👍

xboxfly15 commented:
Thanks for allowing the tests to run, I can see one failed at https://github.com/phpmyadmin/phpmyadmin/actions/runs/12206256316/job/34055909072

xboxfly15 commented:
That's what I get for just copying tests - I believe I should remove https://github.com/phpmyadmin/phpmyadmin/blob/edab453833e4fccbd8e137810ef043c7489d3195/tests/unit/Config/SettingsTest.php#L797C20-L797C63 as type wont be the same

xboxfly15 commented:
Actually there seems to be a lot of tests failing, some unrelated to my code? I can't quite understand them all and if there is anything I should fix. I wont touch anything until you comment.

I pretty much copied `CaptchaSiteVerifyURL`'s tests so not sure why it is failing.

williamdes commented on docs/config.rst:
```diff
@@ -2158,12 +2158,28 @@ Cookie authentication options
+
+    reCaptcha will be then used in :ref:`cookie`.
+
+    .. versionadded:: 5.2.2
```
```suggestion
    .. versionadded:: 6.0.0
```
Due to the target branch of this PR, this is the version it will be released in. Sorry for the first replace
Was it expected ?

williamdes commented on tests/unit/Config/SettingsTest.php:
```diff
@@ -791,6 +794,7 @@ public static function providerForTestConstructor(): array
                     ['CaptchaLoginPublicKey', 1234, '1234'],
                     ['CaptchaLoginPrivateKey', 1234, '1234'],
                     ['CaptchaSiteVerifyURL', 1234, '1234'],
+                    ['captchaExpectedHostname', 1234, '1234'],
```
I think this is the one that makes the test fail, that said I can not make sense of it
@MauricioFauth will probably know why it fails and not CaptchaSiteVerifyURL above

xboxfly15 replied:
I believe so too but also cannot make sense of it and why CaptchaSiteVerifyURL doesn't fail.

xboxfly15 committed:
Update docs/config.rst to new version added

Signed-off-by: xboxfly15 <19935166+xboxfly15@users.noreply.github.com>

Co-authored-by: William Desportes <williamdes@wdes.fr>

williamdes reviewed (approved):
I approve this so it will makes this PR pop up on my merge queue

I will have to rebase your work, thank you for your work and contribution @xboxfly15 !
